pipelines/broad/rna_seq/UMIAwareDuplicateMarking.wdl

```diff
@@ -0,0 +1,148 @@
+      sort_order = "coordinate"
+  }
+
+  # Further divide each duplicate set (a set of reads with the same insert start and end coordinates)
+  # into subsets that share the same UMIs i.e. differenciate PCR duplicates from biological duplicates.
+  # (biological duplicates are independent DNA molecules that are sheared such that the inserts are indistinguishable.)
+  # input: a coordinate sorted bam
+  # output: a coordinate sorted bam with UMIs (what are the generated tags?) .
+  
+  call GroupByUMIs {
+    input:
+      bam = SortSamFirst.output_bam,
+      bam_index = select_first([SortSamFirst.output_bam_index, "bam_index_not_found"]),
+      output_bam_basename = output_basename + ".grouped_by_UMI"
+  }
+
+  # input:
+  # output: 
+  call SortSam as SortSamQueryName {
+    input:
+      input_bam = GroupByUMIs.grouped_bam,
+      output_bam_basename = output_basename + ".grouped.queryname_sorted",
+      sort_order = "queryname"
+  }
+
+  call MarkDuplicates {
+    input:
+      bam = SortSamQueryName.output_bam,
+      output_basename = output_basename
+  }
+
+  call SortSam as SortSamSecond {
+    input:
+      input_bam = MarkDuplicates.duplicate_marked_bam,
+      output_bam_basename = output_basename + ".duplicate_marked.coordinate_sorted",
+      sort_order = "coordinate"
+  }
+
+  output {
+    File duplicate_marked_bam = SortSamSecond.output_bam
+    File duplicate_marked_bam_index = select_first([SortSamSecond.output_bam_index, "bam_index_not_found"])
+    File duplicate_metrics = MarkDuplicates.duplicate_metrics
+  }
+}
+
+task MarkDuplicates {
+  input {
+    File bam
+    String output_basename
+  }
+
+  String output_bam_basename = output_basename + ".duplicate_marked"
+
+  Int disk_size = ceil(3 * size(bam, "GB")) + 128
+  command <<<
+    gatk MarkDuplicates -I ~{bam} --READ_ONE_BARCODE_TAG BX -O ~{output_bam_basename}.bam --METRICS_FILE ~{output_basename}.duplicate.metrics --ASSUME_SORT_ORDER queryname
+  >>>
+
+  output {
+    File duplicate_marked_bam = "~{output_bam_basename}.bam"
+    File duplicate_metrics = "~{output_basename}.duplicate.metrics"
+  }
+
+  runtime {
+    docker: "us.gcr.io/broad-gatk/gatk:4.1.9.0"
+    disks: "local-disk " + disk_size + " HDD"
+    memory: "16 GB"
+  }
+}
+
+task SortSam {
+  input {
+    File input_bam
+    String output_bam_basename
+    String sort_order # "queryname" or "coordinate"
+  }
+
+  # SortSam spills to disk a lot more because we are only store 300000 records in RAM now because its faster for our data so it needs
+  # more disk space.  Also it spills to disk in an uncompressed format so we need to account for that with a larger multiplier
+  Float sort_sam_disk_multiplier = 4.0
+  Int disk_size = ceil(sort_sam_disk_multiplier * size(input_bam, "GiB")) + 256
+
+  String extra_args = if sort_order == "coordinate" then "CREATE_INDEX=true" else ""
+
+
+  command {
+    java -Xms8192m -jar /usr/picard/picard.jar \
```

Comment:
For performance should we use samtools for sorting?

Reply:
Resolving this comment for now.  Ticket to follow up on this at: [GL-1878](https://broadinstitute.atlassian.net/browse/GL-1878)